deerlab/dipolarkernel.py

```diff
@@ -224,65 +254,71 @@ def calckernelmatrix(t,r,method,wex,nKnots,g):
     nr = np.size(r)
     nt = np.size(t)  
     K = np.zeros((nt,nr))
-    wr = w0(g)/(r**3)  # rad s^-1
+    ωr = ω0(g)/(r**3)  # Mrad s^-1
 
-    def kernelmatrix_fresnel(t,r):
+    orientationselection = Pθ is not None 
+
+    def elementarykernel_fresnel(t,r):
     #==========================================================================
         """Calculate kernel using Fresnel integrals (fast and accurate)"""
 
        # Calculation using Fresnel integrals
-        ph = np.outer(np.abs(t), wr)
-        kappa = np.sqrt(6*ph/pi)
+        ɸ = np.outer(np.abs(t), ωr)
+        κ = np.sqrt(6*ɸ/pi)
         
         # Supress divide by 0 warning        
         with warnings.catch_warnings():
             warnings.simplefilter('ignore')
-            S, C = fresnel(kappa)/kappa
+            S, C = fresnel(κ)/κ
         
-        K = C*np.cos(ph) + S*np.sin(ph)
+        K = C*np.cos(ɸ) + S*np.sin(ɸ)
         K[t==0] = 1 
 
         return K
     #==========================================================================
-    
-    def kernelmatrix_grid(t,r,wex,nKnots):
+
+    def elementarykernel_grid(t,r,ωex,nKnots,Pθ):
     #==========================================================================
         """Calculate kernel using grid-based powder integration (converges very slowly with nKnots)"""
 
         # Orientational grid
-        costheta = np.linspace(0,1,int(nKnots))
-        q = 1 - 3*costheta**2
-        # Vectorized 3D-grid evaluation (t,r,powder averaging)
-        C = np.cos(wr[np.newaxis,:,np.newaxis]*q[np.newaxis,np.newaxis,:]*abs(t[:,np.newaxis,np.newaxis]))
-        # If given, include limited excitation bandwidth
-        if not np.isinf(wex):
-            C = C*np.exp(-(wr[np.newaxis,:,np.newaxis]*q[np.newaxis,np.newaxis,:])**2/wex**2)
-        K = np.sum(C,2)/nKnots
+        cosθ = np.linspace(0,1,int(nKnots))
+        q = 1 - 3*cosθ**2
+
+        if orientationselection:
+            # Interpolate
+            Pθ = interp1d(np.linspace(0,1,int(len(Pθ))),Pθ,kind='cubic')(cosθ)
+            Pθ /= np.sum(Pθ)
+            # Integrate over the orientations distribution Pθ
+            K = np.dot(_Cgrid(ωr,t,ωex,q),Pθ)
+        else: 
+            # Elementary kernel without orientation selection
+            K = np.sum(_Cgrid(ωr,t,ωex,q),2)/nKnots
```

Comment:
Why is the 2 there? If P0 is isotropic, then it should be 1, right? Or 1/4pi or similar? Check normalization.

Reply:
The 2 is the summation axis, in python meaning the third axis, i.e. the dimension along `theta`